Bump clang version to 19

danakj opened the pull request:
The version of clangd/clang-tidy on developer machines has slowly diverged from the one on the CI builders, which is causing a slowly increasing amount of pain as clang-tidy CI runs fail (incorrectly) over things that a newer clangd/clang-tidy was perfectly fine with locally. This bumps the Clang version used in the ubuntu builders to 19, which is the most recent in Debian stable.

We use https://apt.llvm.org instead of LLVM's GitHub releases (https://github.com/llvm/llvm-project/releases) as the former more reliably has packages for newer Clang/LLVM versions on x64. The community-build releases binaries on LLVM's GitHub have stopped including Ubuntu packages that match the GitHub x64 Ubuntu workers for some time (for at least the 18 and 19 releases).

By moving to apt.llvm.org packages we only download and install the headers and libraries needed for development, rather than every output of building llvm, which is much faster and saves lots of disk space. We also remove the system installations of other versions of clang/llvm so we should end up using negative disk space. We can no longer easily cache the installation but apt.llvm.org is a reliable end point.

We bump the ubuntu image version for the github workers to 24.04, as apt.llvm.org has stopped building images for 22.10 in 2022 at its end of life.

The `pre_commit` workflow disabled sudo unlike the other workflows that install Clang/LLVM, including the `clang-tidy` workflow (which is also run on `pull_request`). We bring it into alignment with the other workflows so that we can install the llvm packages. And we lock its ubuntu image to 24.04 so that it can be moved in lockstep with the other workflows that depend on Clang/LLVM.

danakj committed:
Bump clang version to 18

Use apt.llvm.org instead of GitHub releases
(https://github.com/llvm/llvm-project/releases) as the former more
reliably has packages for newer clang versions on x64. The releases
binaries have stopped including ubuntu packages that match the GitHub
x64 ubuntu workers for some time (for at least the 18 and 19 releases).

By moving to apt.llvm.org packages we only download and install the
headers and libraries needed for development, rather than every output
of building llvm, which is much faster and saves lots of disk space. We
also remove the system installations of other versions of clang/llvm so
we should end up using negative disk space. We can no longer easily
cache the installation but it wouldn't give us much anyway (unless
apt.llvm.org was down, but it is reliable).

We bump the ubuntu image version for the github workers to 24.10, as
apt.llvm.org has stopped building images for 22.10 in 2022 at its end of
life.

danakj commented:
If this looks good, I believe we will have to change the required workflows in the repository settings (cc: @jonmeow)

jonmeow commented on .github/actions/build-setup-ubuntu/action.yml:
```diff
@@ -22,48 +22,47 @@ runs:
-        ls "${LLVM_PATH}/bin"
+        echo '*** Removing system clang packages'
+        for version in ${installed_clang_version//,/ }; do
+          sudo apt-get remove clang-${version} \
```
Maybe indent the `clang` package with the others, seems easy to overlook here:
```suggestion
          sudo apt-get remove \
            clang-${version} \
```

danakj replied:
Done

jonmeow replied:
Missed?

danakj replied:
Er, yes. But will try to avoid removing anything as well.

danakj replied:
This is all gone now, we just leave what's installed alone.

jonmeow commented on .github/actions/build-setup-ubuntu/action.yml:
```diff
@@ -22,48 +22,47 @@ runs:
-        mv "$LLVM_RELEASE" llvm
-        echo "*** Testing `clang++ --version`"
-        ~/llvm/bin/clang++ --version
+        dpkg -l|grep -E 'clang|llvm'
```
Suggesting spacesto separate commands:

```suggestion
        dpkg -l | grep -E 'clang|llvm'
```

jonmeow replied:
Sorry, better suggestion:
```suggestion
        dpkg -l '*clang*' '*llvm*'
```

danakj replied:
Oh that is nicer, thanks

jonmeow commented on .github/actions/build-setup-ubuntu/action.yml:
```diff
@@ -22,48 +22,47 @@ runs:
-        echo '*** ls "${LLVM_PATH}/bin"'
-        ls "${LLVM_PATH}/bin"
+        echo '*** Removing system clang packages'
+        for version in ${installed_clang_version//,/ }; do
```
I assume 18 needs to be removed in order to avoid conflicts, but why do older versions need to be removed? Maybe add a comment explaining?

danakj replied:
We actually don't need to remove it, the packages are older so they will be upgraded. That does save 15 seconds or so, so why not.

I have rephrased this as `remove_clang_versions`.

In theory we could just use the system 18 packages then, but that approach prevents us from getting the latest patch level fixes, or from outpacing the system image on the workers.

jonmeow commented on .github/actions/build-setup-ubuntu/action.yml:
```diff
@@ -22,48 +22,47 @@ runs:
+            libclang-common-${version}-dev \
+            libclang-${version}-dev \
+            libclang-cpp${version}-dev \
+            libunwind-${version}-dev
```
Perhaps sort this list of packages? It might be easier to scan for entries in lexicographic order

danakj replied:
Done

jonmeow replied:
Maybe `lib` before `lld`?

danakj replied:
Oh, yes :)

jonmeow commented on .github/actions/build-setup-ubuntu/action.yml:
```diff
@@ -22,48 +22,47 @@ runs:
-        ls "${LLVM_PATH}"
-        echo '*** ls "${LLVM_PATH}/bin"'
-        ls "${LLVM_PATH}/bin"
+        echo '*** Removing system clang packages'
```
Was thinking about this, had you considered something like:

```
sudo apt-get remove '^(clang|clangd|libc++|libc++abi|libclang|libomp|libunwind|lld|lldb|llvm)-'
```

i.e., use regex to handle everything in one go, although it prints "not installed" for things

jonmeow replied:
more succinctly:

```
sudo apt-get remove '^(clangd?|lib(c\+\+(abi)?|clang|omp|unwind)|lld|lldb|llvm)-'
```

danakj replied:
This ends up removing way more things, too many things, and breaks stuff:
```
bpfcc-tools bpftrace clang-16 clang-17 clang-18 clang-format-16
  clang-format-17 clang-format-18 clang-tidy-16 clang-tidy-17 clang-tidy-18
  clang-tools-16 clang-tools-17 clang-tools-18 libbpfcc libclang-common-16-dev
  libclang-common-17-dev libclang-common-18-dev libclang-cpp16t64
  libclang-cpp17t64 libclang-cpp18 libclang-rt-16-dev libclang-rt-17-dev
  libclang-rt-18-dev libclang1-16t64 libclang1-17t64 libclang1-18
  libgstreamer1.0-0 liblldb-16t64 liblldb-17t64 liblldb-18 libunwind8
  linux-azure linux-azure-6.11-tools-6.11.0-1013 linux-tools-6.11.0-1013-azure
  linux-tools-azure lld-16 lld-17 lld-18 lldb-18 llvm-16 llvm-16-dev
  llvm-16-linker-tools llvm-16-runtime llvm-16-tools llvm-17 llvm-17-dev
  llvm-17-linker-tools llvm-17-runtime llvm-17-tools llvm-18 llvm-18-dev
  llvm-18-linker-tools llvm-18-runtime llvm-18-tools packagekit
  packagekit-tools python3-bpfcc python3-lldb-18 software-properties-common
  strace ubuntu-server ubuntu-standard xvfb
```

We just need to remove enough stuff that we can be sure we'll use the correct binaries and libraries.

jonmeow reviewed (commented):
Note, this PR currently only bumps the ubuntu version. Is that intended?

macos is a separate directory, and probably easier to update; mentions of clang/llvm-16 in contribution_tools.md should probably also be updated to reflect a new minimum if it's changing.

danakj commented:
> Note, this PR currently only bumps the ubuntu version. Is that intended?
> 
> macos is a separate directory, and probably easier to update; mentions of clang/llvm-16 in contribution_tools.md should probably also be updated to reflect a new minimum if it's changing.

The clang version is also updated, see here: https://github.com/carbon-language/carbon-lang/pull/5440/files#diff-f7f6204c4fa72789a0029dfe30d166ffd31488c91bafcaeffd94dd9522223687R36

That value is given to the apt.llvm.org script to install clang 18 here: https://github.com/carbon-language/carbon-lang/pull/5440/files#diff-f7f6204c4fa72789a0029dfe30d166ffd31488c91bafcaeffd94dd9522223687R68

danakj commented:
> > Note, this PR currently only bumps the ubuntu version. Is that intended?
> > macos is a separate directory, and probably easier to update; mentions of clang/llvm-16 in contribution_tools.md should probably also be updated to reflect a new minimum if it's changing.
> 
> The clang version is also updated, see here: https://github.com/carbon-language/carbon-lang/pull/5440/files#diff-f7f6204c4fa72789a0029dfe30d166ffd31488c91bafcaeffd94dd9522223687R36
> 
> That value is given to the apt.llvm.org script to install clang 18 here: https://github.com/carbon-language/carbon-lang/pull/5440/files#diff-f7f6204c4fa72789a0029dfe30d166ffd31488c91bafcaeffd94dd9522223687R68

Oh I get what you mean.. it's bumping the ubuntu version _of clang_.

Let me look at macos, I hadn't thought about that.

jonmeow commented:
FWIW, I think the macos update should be more straightforward version number replacement.

danakj committed:
bump-macos-to-18

danakj committed:
better dpkg command

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

jonmeow commented:
FYI you might want to add a version number to the cache name (it's probably reusing the `@16` cache)

danakj committed:
concise

danakj committed:
add-cache-name-number

danakj committed:
less-concise

danakj committed:
move-macos-cache-name

danakj committed:
move-macos-cache-name-again

danakj commented:
> mentions of clang/llvm-16 in contribution_tools.md should probably also be updated to reflect a new minimum if it's changing.

Thanks for pointing those out, I've updated them to 18, and tried to find any other references to 16 in our bazel which no longer apply.

danakj commented:
> > mentions of clang/llvm-16 in contribution_tools.md should probably also be updated to reflect a new minimum if it's changing.
> 
> Thanks for pointing those out, I've updated them to 18, and tried to find any other references to 16 in our bazel which no longer apply.

I've done something bad to the lockfile. It claims to need updating so I did `bazel mod deps --lockfile_mode=update` which made a small change, but then building anything makes more changes, and none of these seem to be okay on the bots.

jonmeow commented:
> > > mentions of clang/llvm-16 in contribution_tools.md should probably also be updated to reflect a new minimum if it's changing.
> > 
> > 
> > Thanks for pointing those out, I've updated them to 18, and tried to find any other references to 16 in our bazel which no longer apply.
> 
> I've done something bad to the lockfile. It claims to need updating so I did `bazel mod deps --lockfile_mode=update` which made a small change, but then building anything makes more changes, and none of these seem to be okay on the bots.

At a glance, maybe you need to merge in https://github.com/carbon-language/carbon-lang/pull/5445 ? Note lockfile versions changed in that PR

jonmeow commented on docs/project/contribution_tools.md:
```diff
@@ -88,25 +88,28 @@ it to your `$PATH`, and aliasing `bazel` to it.
 
-> NOTE: Most LLVM 16+ installs should build Carbon. If you're having issues, see
+Alternatively, you can install Clang tools on Debian-based distros from
+[https://apt.llvm.org](https://apt.llvm.org).
```
```suggestion
<https://apt.llvm.org>.
```

danakj replied:
Done

jonmeow commented on docs/project/contribution_tools.md:
````diff
@@ -88,25 +88,28 @@ it to your `$PATH`, and aliasing `bazel` to it.
 ```
 
-> NOTE: Most LLVM 16+ installs should build Carbon. If you're having issues, see
+Alternatively, you can install Clang tools on Debian-based distros from
````
Had you considered phrasing this as more of a fallback, instead of a simple alternative? I've avoided mentioning it previously because most users shouldn't (typically) need it, and it won't work on all Debian/Ubuntu distributions. For example, your work machine should be a Debian-based distribution: have you tried installing from apt.llvm.org?

But also, apt.llvm.org says "Debian/Ubuntu"; this would mention how we talk about "Debian or Ubuntu" above. Even if we mention it, would it make sense to keep that way of referring to Linux distros? Note this may be a real issue for some: apt.llvm.org assumes it can map to a specific Debian or Ubuntu version, which may not work with arbitrary Debian-based distributions.

Note, this had me looking at how others talk about Linux. https://redhat-documentation.github.io/supplementary-style-guide/ says "Linux distributions" only, never "distro". https://wiki.ubuntu.com/DocumentationTeam/StyleGuide/StandardTerminology says to avoid saying "distros" but define it if used. That's a bit why I'm suggesting just copying what apt.llvm.org says for a mention.

jonmeow replied:
Actually, do we even need it for GitHub if llvm-18 is installed by default? I think there may need to ensure the right libstdc++ is installed, but maybe we could use the same apt command (as what we recommend on this page)? I know in the past there was a difference in build flags (I think for fpic issues), and maybe that's been fixed.

danakj replied:
After open discussion we are going to go right to 19 here. I've changed the wording as you've suggested, thanks.

jonmeow commented on .github/actions/build-setup-ubuntu/action.yml:
```diff
@@ -22,48 +22,45 @@ runs:
+        clang_version: 18
+        # The ubuntu 24.04 image has 16, 17, and 18 installed.
+        # https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md
+        # We r
```
Unfinished comment?

danakj replied:
Done

jonmeow reviewed (commented):
Thanks, I think this is covering everything I'm aware of -- I'd forgotten the cc_toolchains stuff.

Probably still good to make sure Chandler's happy to increase the minimum version (I'm not sure if he's weighed in already somewhere)

danakj committed:
bump-min-version

danakj committed:
19

danakj commented on .github/actions/build-setup-ubuntu/action.yml:
```diff
@@ -22,48 +22,27 @@ runs:
+        chmod +x llvm.sh
+        sudo ./llvm.sh ${clang_version} all
+
+        sudo ln -sf /lib/llvm-19/bin/clang /bin/
+        sudo ln -sf /lib/llvm-19/bin/clang++ /bin/
+        sudo ln -sf /lib/llvm-19/bin/clang-format /bin/
+        sudo ln -sf /lib/llvm-19/bin/clang-tidy /bin/
+        sudo ln -sf /lib/llvm-19/bin/lld /bin/
+        sudo ln -sf /lib/llvm-19/bin/llvm-symbolizer /bin/
```
I could not determine a way to set `CC` globally, and having to set it in every action is pretty flaky. But this works globally so that all future actions see the right thing, which we nicely print out with `which` and `--version` queries. If I just do nothing, it still tries to use 18.

chandlerc replied:
Instead, you could add this directory to the `PATH` environment variable?

https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#adding-a-system-path

danakj replied:
Oh yes let me give it a shot, thanks.

danakj replied:
Yep that worked, thank you

danakj commented:
> > > > mentions of clang/llvm-16 in contribution_tools.md should probably also be updated to reflect a new minimum if it's changing.
> > > 
> > > 
> > > Thanks for pointing those out, I've updated them to 18, and tried to find any other references to 16 in our bazel which no longer apply.
> > 
> > 
> > I've done something bad to the lockfile. It claims to need updating so I did `bazel mod deps --lockfile_mode=update` which made a small change, but then building anything makes more changes, and none of these seem to be okay on the bots.
> 
> At a glance, maybe you need to merge in #5445 ? Note lockfile versions changed in that PR

I tried merging it and it still broke a lot. I have reverted the changes to bazel so that things work. We can look at bumping the 16 to 19 in the bazel file separately, as there's too many moving pieces in this PR already. Even adding TODOs changes the digest which breaks stuff.

chandlerc commented on .github/actions/build-setup-ubuntu/action.yml:
```diff
@@ -22,48 +22,27 @@ runs:
       shell: bash
       run: |
-        cd ~
-        LLVM_RELEASE=clang+llvm-16.0.4-x86_64-linux-gnu-ubuntu-22.04
```
FWIW, could we just use the old code and use LLVM-19 releases here? I'm worried we're going to re-discover the problems we hit with the debian packaging before.

danakj replied:
Not unless we build our own packages. There's no ubuntu packages in the community-driven binary releases for x64 since at least 18. This is mentioned in the PR description too.

danakj replied:
Also, just fwiw I used these apt.llvm.org packages on subspace CI for a year without any issues, which is why I knew to suggest it.

chandlerc replied:
Hmm, ok. We had definitely hit issues with `apt.llvm.org` in the past, but let's see if it works better with the newer packages.

FWIW, I suspect the "Linux" package files would work fine here (I don't think we needed Ubuntu-specific files) if we do hit issues.

danakj committed:
path

danakj committed:
nolint-tidy-19

jonmeow reviewed (approved):
To be sure, from my perspective, approved. Thank you for doing this update!

Approving so that if Chandler confirms he's okay, you can merge. :)

chandlerc reviewed (approved):
LG

danakj commented:
@jonmeow we'll need some admin powers to change the required workflows in order to merge this, and then to turn on the new ones after
